Great Expectations Example

samhita-alla opened the pull request:
Signed-off-by: Samhita Alla <aallasamhita@gmail.com>

samhita-alla commented on .github/workflows/ghcr_push.yml:
```diff
@@ -39,6 +39,8 @@ jobs:
             path: integrations/flytekit_plugins
           - name: pandera
             path: integrations/flytekit_plugins
+          - name: greatexpectations
```
There's one `great_expectations` directory within the parent `greatexpectations` which houses the configuration. Hence, I don't think we can change this. Please look at the directory structure: https://github.com/flyteorg/flytesnacks/blob/f4a6e36440ca1fb51f5203858ee10ae2069aa5d2/cookbook/integrations/flytekit_plugins/greatexpectations/great_expectations/expectations/sqlite/movies.json

talagluck replied:
Hi @samhita-alla, I thought I would check if this could be changed, since it looks like a few of the other directories in this list have underscores. I understand if this is not possible.

samhita-alla replied:
Yeah. We can have `great_expectations`, but since there's this config directory `great_expectations` present within it, it may confuse the users. The internal config directory is the one that's directly generated from Great Expectations' CLI.

talagluck replied:
I see - that makes sense. Let me bring this up with the team at our meeting next week to see if anyone has any ideas.

samhita-alla replied:
Sure!

talagluck replied:
Just following up - I brought this up with the team, and the consensus was that if it is possible, we would like to use `great_expectations/great_expectations`. If you think it will be too confusing, we can leave as is. Thanks!

samhita-alla replied:
`greatexpectations` is now `great_expectations`.

samhita-alla replied:
I've reverted this to `greatexpectations` because there's a conflict between `great_expectations` in this directory and the usual `great_expectations` library in the plugin code.

samhita-alla commented on cookbook/integrations/flytekit_plugins/greatexpectations/README.rst:
```diff
@@ -0,0 +1,68 @@
+How to Define Your Integration
+------------------------------
+
+Great Expectations accepts datasets that can only be "strings" -- file names, table names, etc. 
```
Oh, I'm sorry! I was referring to `data_asset`, not the Datasource. Rephrased it now. Let me know if the update looks okay. :)

talagluck replied:
No worries! This has also changed a bit over time. I am still not sure whether this is entirely accurate, but I also may be misunderstanding some things here. 

To clarify: 
If your Datasource is a directory of csvs, or a SQL table, then the statement here is correct - you would specify the csv file or table as a string with an `Inferred` or `ConfiguredAssetDataConnector`, and then the DataConnector would help retrieve the object with that name. 

However, you could also use a RuntimeDataConnector to pass in a pandas or Spark DataFrame, in which case you would be passing the DataFrame object directly into the BatchRequest. In this case, Great Expectations still requires you to specify a `data_asset_name`, but this is essentially an arbitrary string, for users to help keep track of the data used. It is this latter case of RuntimeDataConnectors which I think might not be accurately depicted here.

Am I understanding this correctly? It is totally possible that I might be missing something here!

samhita-alla replied:
@talagluck, I think I got your point. We have the privilege to give a non-string (DataFrame) when using a `RuntimeDataConnector` but not when using an `Inferred` or a `ConfiguredAssetDataConnector` (correct?).

I have a question here. When I was surfing through the examples in Great Expectations' Docs, I've come across this example:

```python
df: pd.DataFrame = pd.read_csv("some_path.csv")
runtime_batch_request = RuntimeBatchRequest(
    datasource_name="my_pandas_datasource",
    data_connector_name="my_runtime_data_connector",
    data_asset_name="insert_your_data_asset_name_here",
    runtime_parameters={
      "batch_data": df
    },
    batch_identifiers={
        "some_key_maybe_pipeline_stage": "ingestion step 1",
        "some_other_key_maybe_airflow_run_id": "run 18"
    }
)
```
(I believe this is what you're referring to)

`runtime_parameters` has `batch_data` which allows us to give a DataFrame object. However, there's this line in the changelog saying that `batch_data` isn't available anymore: https://docs.greatexpectations.io/en/latest/changelog.html?highlight=RuntimeDataConnector#id9

So now, what's the correct way of passing in a DataFrame object? Can you point me to an example if there's any? Also, this is the type I've attached to `runtime_parameters` in the plugin:

```python
runtime_parameters: typing.Optional[typing.Dict[str, str]] = None
```
Should this be modified to `runtime_parameters: typing.Optional[typing.Dict[str, typing.Any]] = None`?

talagluck replied:
Hi @samhita-alla

> We have the privilege to give a non-string (DataFrame) when using a RuntimeDataConnector but not when using an Inferred or a ConfiguredAssetDataConnector (correct?).

This is exactly right.

> runtime_parameters has batch_data which allows us to give a DataFrame object. However, there's this line in the changelog saying that batch_data isn't available anymore: https://docs.greatexpectations.io/en/latest/changelog.html?highlight=RuntimeDataConnector#id9

I understand the confusion here. `batch_data` was removed from the `BatchRequest` object because we added a `RuntimeBatchRequest` object that uses it instead. It appears that your code in the Flyte integration already takes into account the possibility of using a `RuntimeBatchRequest` through the `BatchRequestConfig` object, so I don't know if much will need to be updated there. [You can see an example of using a `RuntimeBatchRequest` in this way here](https://docs.greatexpectations.io/en/latest/guides/how_to_guides/creating_batches/how_to_load_a_pandas_dataframe_as_a_batch.html) under the Show Docs for V3 tab.

> Also, this is the type I've attached to runtime_parameters in the plugin:

> runtime_parameters: typing.Optional[typing.Dict[str, str]] = None

> Should this be modified to runtime_parameters: typing.Optional[typing.Dict[str, typing.Any]] = None?

Yes - this looks right in light of the ability to pass in a DataFrame here.

Thanks for the follow-up here - some great questions!

samhita-alla replied:
@talagluck: Can we not have an automated mechanism in Great Expectations wherein `data_asset_name` is automatically picked by `runtime_parameters` if the batch request is a `RuntimeBatchRequest`? This is because `data_asset_name` doesn't add any value if we're using a `RuntimeBatchRequest`.

(Say) In the case of DataFrames, you can populate the `batch_data`, and in the case of SQL queries, you can populate the table name.

What do you think?

samhita-alla replied:
Okay, thank you!

samhita-alla commented on cookbook/integrations/flytekit_plugins/greatexpectations/task_example.py:
```diff
@@ -0,0 +1,156 @@
+#
+# We define a ``GreatExpectationsTask`` that validates a CSV file. This does pandas data validation.
+simple_task_object = GreatExpectationsTask(
+    name="getask_simple",
```
Sure! 
Modified the task names to have `great_expectations`, but not `ge`.

samhita-alla commented:
@talagluck: Thank you for reviewing the code! I've added runtime batch request-related examples. I've made some changes in the plugin to consider the dataset as the runtime batch data (`data_asset_name` would be a random string in such a case). You can check them out! If everything looks okay, we can proceed with merging the two PRs.

talagluck commented on cookbook/integrations/flytekit_plugins/greatexpectations/README.rst:
```diff
@@ -0,0 +1,111 @@
+  and validate the data using Great Expectations.
+
+.. note::
+  Flyte types are added because, in Great Expectations, we have the privilege to give a non-string (Pandas/Spark DataFrame) when using a 
+  :py:class:`RuntimeDataConnector <greatexpectations:great_expectations.datasource.data_connector.runtime_data_connector.RuntimeDataConnector>` 
+  but not when using an 
+  :py:class:`InferredAssetFilesystemDataConnector <great_expectations.datasource.data_connector.inferred_asset_filesystem_data_connector.InferredAssetFilesystemDataConnector>` 
+  or a 
+  :py:class:`ConfiguredAssetFilesystemDataConnector <great_expectations.datasource.data_connector.configured_asset_filesystem_data_connector.ConfiguredAssetFilesystemDataConnector>`. 
+  For the latter case, with the integration of Flyte types, we can give a Pandas/Spark DataFrame or a remote URI as the dataset.
```
Love this language - thanks for the update!

talagluck commented:
Thanks for the updates here as well, @samhita-alla!

kumare3 commented:
This looks pretty cool to me. +1

samhita-alla commented:
> This looks pretty cool to me. +1

@kumare3 Thank you! Let's merge this after the plugin is merged. Also, I'm adding two dataset files (one CSV and one SQLite) as we have some examples that use local data. Is that okay?

samhita-alla commented:
Hi, @talagluck! When will API reference be updated in the new Great Expectations docs?

samhita-alla commented:
@talagluck Here’s a blog I’ve written that talks about our integration in brief: https://docs.google.com/document/d/1yIOqMnbBUQJn4-_b1LSXm1KU7CRR1qBn1p6QaTCmA-c/edit?usp=sharing. Please feel free to comment.

talagluck commented:
Thanks so much, @samhita-alla! I'll take a look at the blog post doc tomorrow.

As far as the API reference - could you point me to the links in the Flyte reference? I'll comment with the updated links. Sorry for the inconvenience!

samhita-alla commented:
> Thanks so much, @samhita-alla! I'll take a look at the blog post doc tomorrow.
> 
> As far as the API reference - could you point me to the links in the Flyte reference? I'll comment with the updated links. Sorry for the inconvenience!

No problem! Let me point you towards the links.

samhita-alla commented on cookbook/integrations/flytekit_plugins/greatexpectations/README.rst:
```diff
@@ -0,0 +1,111 @@
+  and validate the data using Great Expectations.
+
+.. note::
+  Flyte types are added because, in Great Expectations, we have the privilege to give a non-string (Pandas/Spark DataFrame) when using a 
+  :py:class:`RuntimeDataConnector <greatexpectations:great_expectations.datasource.data_connector.runtime_data_connector.RuntimeDataConnector>` 
+  but not when using an 
+  :py:class:`InferredAssetFilesystemDataConnector <great_expectations.datasource.data_connector.inferred_asset_filesystem_data_connector.InferredAssetFilesystemDataConnector>` 
+  or a 
+  :py:class:`ConfiguredAssetFilesystemDataConnector <great_expectations.datasource.data_connector.configured_asset_filesystem_data_connector.ConfiguredAssetFilesystemDataConnector>`. 
```
We have to update links here.

talagluck replied:
https://legacy.docs.greatexpectations.io/en/latest/autoapi/great_expectations/datasource/data_connector/configured_asset_filesystem_data_connector/index.html#great_expectations.datasource.data_connector.configured_asset_filesystem_data_connector.ConfiguredAssetFilesystemDataConnector

samhita-alla commented on cookbook/integrations/flytekit_plugins/greatexpectations/README.rst:
```diff
@@ -0,0 +1,111 @@
+^^^^^^^^^^^^^^^^^^^^^^^
+
+If the data validation fails, the plugin will raise a Great Expectations' 
+:py:obj:`ValidationError <great_expectations.exceptions.ValidationError>`.
```
...

talagluck replied:
https://legacy.docs.greatexpectations.io/en/latest/autoapi/great_expectations/exceptions/index.html#great_expectations.exceptions.ValidationError

samhita-alla commented on cookbook/integrations/flytekit_plugins/greatexpectations/README.rst:
```diff
@@ -0,0 +1,111 @@
+^^^^^^^^^^^^^^^^^^^
+
+- **context_root_dir**: Sets the path of the great expectations config directory. 
+- **checkpoint_params**: Optional :py:class:`greatexpectations:great_expectations.checkpoint.checkpoint.SimpleCheckpoint` class parameters.
```
...

talagluck replied:
https://legacy.docs.greatexpectations.io/en/latest/autoapi/great_expectations/checkpoint/checkpoint/index.html#great_expectations.checkpoint.checkpoint.SimpleCheckpoint

samhita-alla commented on cookbook/integrations/flytekit_plugins/greatexpectations/task_example.py:
```diff
@@ -0,0 +1,209 @@
+# RuntimeBatchRequest
+# ===================
+#
+# The :py:class:`RuntimeBatchRequest <great_expectations.core.batch.RuntimeBatchRequest>` can wrap either an in-memory DataFrame,
```
...

talagluck replied:
https://legacy.docs.greatexpectations.io/en/latest/autoapi/great_expectations/core/batch/index.html#great_expectations.core.batch.RuntimeBatchRequest

talagluck reviewed (commented):
Hi @samhita-alla ! Thanks for pointing out the broken links. I've added links to our legacy docs for these. We are still finalizing this part for our new docs migration, so we should be able to update these in the near future, but these links should be fine for now.

samhita-alla commented:
> Hi @samhita-alla ! Thanks for pointing out the broken links. I've added links to our legacy docs for these. We are still finalizing this part for our new docs migration, so we should be able to update these in the near future, but these links should be fine for now.

Thank you, @talagluck! Updated the links now.

samhita-alla commented:
Let's merge this PR after the https://github.com/great-expectations/great_expectations/issues/3290 issue is fixed.
